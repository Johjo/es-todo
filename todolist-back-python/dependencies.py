from dataclasses import dataclass, field, replace
from enum import Enum
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)



    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        assert (resource_type, resource) in self.factory, f"{resource_type.value} for {resource} must be injected first"
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

