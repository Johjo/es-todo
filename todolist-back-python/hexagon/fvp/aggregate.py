from abc import ABC, abstractmethod
from collections import OrderedDict
from dataclasses import dataclass

from hexagon.shared.type import TaskKey

@dataclass(frozen=True)
class Task:
    key: TaskKey

    def to_do_the_task(self):
        return DoTheTask(key=self.key)

    def to_choose_the_task(self, other_task: 'Task'):
        return ChooseTheTask(main_task_key=self.key, secondary_task_key=other_task.key)


@dataclass(frozen=True)
class NothingToDo:
    pass


@dataclass(frozen=True)
class DoTheTask:
    key: TaskKey


@dataclass(frozen=True)
class ChooseTheTask:
    main_task_key: TaskKey
    secondary_task_key: TaskKey


@dataclass(frozen=True)
class FvpSnapshot:
    task_priorities: OrderedDict[TaskKey, TaskKey]

    def to_primitive_dict(self) -> dict[TaskKey, TaskKey]:
        return {key: value for key, value in self.task_priorities.items()}

    @classmethod
    def from_primitive_dict(cls, data: dict[TaskKey, TaskKey]) -> 'FvpSnapshot':
        return FvpSnapshot(OrderedDict[TaskKey, TaskKey](data))


class FinalVersionPerfectedSession:
    def __init__(self, task_priorities: OrderedDict[TaskKey, TaskKey]):
        self.task_priorities: OrderedDict[TaskKey, TaskKey] = task_priorities

    def which_task(self, open_tasks: list[Task]) -> NothingToDo | DoTheTask | ChooseTheTask:
        ids = [task.key for task in open_tasks]
        tasks = [task for task in open_tasks if self.task_priorities.get(task.key, None) not in ids]
        match tasks:
            case [task]:
                return task.to_do_the_task()
            case [task_1, task_2, *tasks]:
                return task_1.to_choose_the_task(task_2)
            case _:
                return NothingToDo()

    def choose_and_ignore_task(self, id_chosen: TaskKey, id_ignored: TaskKey) -> None:
        self.task_priorities[id_ignored] = id_chosen

    def reset(self) -> None:
        self.task_priorities = OrderedDict()

    def to_snapshot(self) -> FvpSnapshot:
        return FvpSnapshot(self.task_priorities)

    @classmethod
    def from_snapshot(cls, snapshot: FvpSnapshot) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(task_priorities=snapshot.task_priorities)

    @classmethod
    def create(cls) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(OrderedDict())

    def cancel_priority(self, task_key: TaskKey):
        keys = list()
        for (key, value) in self.task_priorities.items():
            if value == task_key:
                keys.append(key)

        for key in keys:
            del self.task_priorities[key]


class FvpSessionSetPort(ABC):
    @abstractmethod
    def save(self, snapshot: FvpSnapshot) -> None:
        pass

    @abstractmethod
    def by(self) -> FvpSnapshot:
        pass
