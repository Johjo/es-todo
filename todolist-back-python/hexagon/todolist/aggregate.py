from dataclasses import dataclass, replace

from expression import Result, Error, Ok


@dataclass(frozen=True)
class TaskKey:
    value: int


@dataclass
class TaskSnapshot:
    key: TaskKey
    name: str
    is_open: bool


@dataclass
class TodolistSnapshot:
    name: str
    tasks: list[TaskSnapshot]


@dataclass(frozen=True)
class Task:
    key: TaskKey
    name: str
    is_open: bool

    def to_snapshot(self) -> TaskSnapshot:
        return TaskSnapshot(key=self.key, name=self.name, is_open=self.is_open)

    @classmethod
    def from_snapshot(cls, snapshot: TaskSnapshot) -> 'Task':
        return Task(key=snapshot.key, name=snapshot.name, is_open=snapshot.is_open)

    def reword(self, new_name) -> 'Task':
        return replace(self, name=new_name)

    def close_task(self) -> 'Task':
        return replace(self, is_open=False)


@dataclass(frozen=True)
class TodolistAggregate:
    name: str
    tasks: tuple[Task, ...]

    @classmethod
    def create(cls, todolist_name) -> 'TodolistAggregate':
        return TodolistAggregate(name=todolist_name, tasks=())

    @classmethod
    def from_snapshot(cls, snapshot: TodolistSnapshot) -> 'TodolistAggregate':
        return TodolistAggregate(name=snapshot.name, tasks=(*[Task.from_snapshot(task) for task in snapshot.tasks],))

    def to_snapshot(self) -> TodolistSnapshot:
        return TodolistSnapshot(self.name, tasks=[task.to_snapshot() for task in self.tasks])

    def open_task(self, task: Task) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (task,)))

    def close_task(self, key) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")
        return Ok(replace(self, tasks=(*[task.close_task() if task.key == key else task for task in self.tasks],)))

    def reword_task(self, key: TaskKey, new_name: str) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        return Ok(replace(self, tasks=(*[task.reword(new_name) if task.key == key else task for task in self.tasks],)))

    def import_tasks(self, task_snapshots: list[TaskSnapshot]) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (*[Task.from_snapshot(snapshot) for snapshot in task_snapshots],)))

