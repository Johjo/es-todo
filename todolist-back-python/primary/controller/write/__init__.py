from primary.controller.write.add_task import add_task
from primary.controller.write.choose_and_ignore_task import choose_and_ignore_task
from primary.controller.write.close_task import close_task
from primary.controller.write.create_todolist import create_todolist
from primary.controller.write.import_todolist_from_markdown import import_todolist_from_markdown
from primary.controller.write.reset_fvp_algorithm import reset_fvp_algorithm
from primary.controller.write.reword_task import reword_task




