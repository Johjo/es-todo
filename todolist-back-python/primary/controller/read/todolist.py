from abc import ABC, abstractmethod
from dataclasses import dataclass

from dependencies import Dependencies
from hexagon.shared.type import TaskKey, TodolistName, TodolistContext, TodolistContextCount, TaskOpen, TaskName


# todo : use a task key instead of an id
@dataclass
class Task:
    id: TaskKey
    name: TaskName
    is_open: TaskOpen

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name
        }


class TodolistSetReadPort(ABC):
    @abstractmethod
    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        pass

    @abstractmethod
    def all_by_name(self) -> list[TodolistName]:
        pass

    @abstractmethod
    def counts_by_context(self, todolist_name: TodolistName) -> list[tuple[TodolistContext, TodolistContextCount]]:
        pass

    @abstractmethod
    def all_tasks(self, todolist_name: TodolistName) -> list[Task]:
        pass


class TodolistReadController:
    def __init__(self, dependencies: Dependencies):
        self.dependencies = dependencies

    def all_todolist_by_name(self) -> list[str]:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.all_by_name()

    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.task_by(todolist_name=todolist_name, task_key=task_key)

    def counts_by_context(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.counts_by_context(todolist_name=todolist_name)

    def export_to_markdown(self):
        return ""

    def to_markdown(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return to_markdown(todolist_set.all_tasks(todolist_name))


def to_markdown(tasks: list[Task]) -> str:
    def task_to_markdown(task: Task) -> str:
        return f"- [{" " if task.is_open else "x"}] {task.name}"
    return "\n".join([task_to_markdown(task) for task in tasks])
