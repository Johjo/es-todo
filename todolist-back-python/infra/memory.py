from uuid import UUID

from expression import Option, Nothing, Some

from hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[str, TodolistSnapshot] = {}

    def by(self, todolist_name: str) -> Option[TodolistSnapshot]:
        if todolist_name not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[todolist_name])

    def save(self, todolist: TodolistSnapshot):
        self.all_todolist[todolist.name] = todolist

    def task_by(self, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[todolist_name].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self) -> list[str]:
        return [name for name in self.all_todolist.keys()]

    def all_tasks(self, todolist_name: str) -> list[TaskSnapshot]:
        if todolist_name not in self.all_todolist:
            return []
        return list(self.all_todolist[todolist_name].tasks)


