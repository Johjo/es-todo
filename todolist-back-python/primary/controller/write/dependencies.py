from abc import ABC
from collections import namedtuple
from dataclasses import dataclass, replace, field
from typing import Any

from hexagon.todolist.aggregate import TodolistSetPort
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.open_task import OpenTask


@dataclass(frozen=True)
class Dependencies(ABC):
    use_case_factory: dict[Any, Any] = field(default_factory=dict)
    adapter_factory: Any = None

    def get_adapter(self, port) -> Any:
        assert self.adapter_factory, f"adapter for {port} must be injected first"
        return self.adapter_factory(self)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return replace(self, use_case_factory={**self.use_case_factory, use_case: use_case_factory})


    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return replace(self, adapter_factory=adapter_factory)

    def get_use_case(self, use_case: Any) -> Any:
        assert use_case in self.use_case_factory, f"use_case for {use_case} must be injected first"
        return self.use_case_factory[use_case](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    factories = {
        TodolistCreate: todolist_create_factory,
        OpenTask : open_task_use_case_factory,
    }

    for use_case, factory in factories.items():
        dependencies = dependencies.feed_use_case(use_case, factory)
    return dependencies


def todolist_create_factory(dependencies: Dependencies):
    return TodolistCreate(dependencies.get_adapter(TodolistSetPort))

def open_task_use_case_factory(dependencies: Dependencies):
    return OpenTask(dependencies.get_adapter(TodolistSetPort))
