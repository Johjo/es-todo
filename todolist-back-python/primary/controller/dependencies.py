from functools import reduce
from typing import Any

from dependencies import Dependencies
from hexagon.fvp.read.which_task import WhichTaskQuery
from hexagon.fvp.write.cancel_priority import CancelPriority
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    use_cases_with_factory: list[Any] = [ResetFvpSession, CancelPriority, ChooseAndIgnoreTaskFvp, TodolistCreate,
                                         OpenTaskUseCase, CloseTask, RewordTask, ImportManyTask, WhichTaskQuery]

    def feed_use_case(dep: Dependencies, use_case: Any) -> Dependencies:
        return dep.feed_use_case(use_case=use_case, use_case_factory=use_case.factory)

    return reduce(feed_use_case, use_cases_with_factory, dependencies)


