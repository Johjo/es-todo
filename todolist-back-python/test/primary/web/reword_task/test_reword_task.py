from dataclasses import replace

from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter  # type: ignore
from webtest import TestApp  # type: ignore

from hexagon.todolist.aggregate import TaskSnapshot
from dependencies import Dependencies
from primary.controller.write.todolist import TodolistWriteController
from primary.web.pages import bottle_config
from test.hexagon.todolist.fixture import TodolistFaker, a_task_key, TodolistSetForTest
from test.hexagon.todolist.write.test_open_task import TaskKeyGeneratorForTest


def test_reword_task(todolist_set: TodolistSetForTest, task_key_generator : TaskKeyGeneratorForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies
    initial_task = fake.a_task(1)
    expected_task = replace(initial_task, name="reworded")

    todolist = replace(fake.a_todolist(), name="todolist", tasks=[initial_task])
    todolist_set.feed(todolist)

    response = app.post(f'/todo/{todolist.name}/item/{expected_task.key.value}/reword', {"new_name": expected_task.name})

    assert response.status == '200 OK'
    assert expected_task in todolist_set.by(todolist.name).value.tasks
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())

