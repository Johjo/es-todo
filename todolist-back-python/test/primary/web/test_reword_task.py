from webtest import TestApp  # type: ignore

from src.dependencies import Dependencies
from src.infra.memory import Memory
from src.primary.web.pages import bottle_config
from test.fixture import TodolistFaker
from test.hexagon.todolist.fixture import TaskKeyGeneratorForTest
from test.primary.web.fixture import CleanResponse, BASE_URL, header_with_good_authentication


def test_reword_task(memory: Memory, task_key_generator : TaskKeyGeneratorForTest, dependencies: Dependencies, app: TestApp, fake: TodolistFaker) -> None:
    # given
    bottle_config.dependencies = dependencies
    initial_task = fake.a_task()
    expected_task = initial_task.having(name="reworded")

    todolist = fake.a_todolist().having(tasks=[initial_task])
    memory.save(user_key="test@mail.fr", todolist=todolist.to_snapshot())

    # when
    response = app.post(f'{BASE_URL}/{todolist.to_name()}/item/{expected_task.to_key()}/reword', {"new_name": expected_task.to_name()}, headers=header_with_good_authentication())

    # then
    assert expected_task.to_snapshot() in memory.by(user_key="test@mail.fr", todolist_name=todolist.to_name()).value.tasks
    assert CleanResponse(response).location() == f"/todo/{todolist.to_name()}"
    assert response.status_code == 302

