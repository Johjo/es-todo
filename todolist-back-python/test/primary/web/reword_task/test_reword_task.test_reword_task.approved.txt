b'<html>

<head>

    <meta charset="UTF-8">

    <title>Todo</title>

</head>

<body>



<p>

    <div>

    <h2> reworded </h2> <a href="/todo/todolist/item/00000000-0000-0000-0000-000000000001/reword?xxxx_query_string">Renommer</a>

    </div>

    <form action="/todo/todolist/item/00000000-0000-0000-0000-000000000001/close?xxxx_query_string" method="post">

    <input type="submit" value="C\'est fait">

    </form>

</p>







<div>

Ajouter une t\xc3\xa2che

<form action="/todo/todolist/item?xxxx_query_string" method="post">

    <input type="text" name="task_name">

    <input type="submit" value="Add item">

</form>

</div>



Red\xc3\xa9marrer l\'algo

<form action="/todo/todolist/reset?xxxx_query_string" method="post">

    <input type="submit" value="Reset">

</form>



Il y a xxxx_number_of_items t\xc3\xa2ches en attente.





<h2> Contexte </h2>

<ul>

<li> <a href="/todo/todolist?only_inbox=1">Inbox</a></li>

<li> <a href="/todo/todolist?context=xxxx1">xxxx1</a> : xxxx2 </li>

</ul>





<ul>

<li><a href="/todo/todolist/export">Exporter les t\xc3\xa2ches</a></li>

<li><a href="/todo/todolist/import">Importer des t\xc3\xa2ches</a></li>

</ul>







</body>

</html>'
